File: doe_modules/simulation/dunnett_power.R
### dunnett_power.R ###

options(warn = -1)

# packages <- c("mvtnorm", "stats")

# for (pkg in packages) {
#   if (!requireNamespace(pkg, quietly = TRUE)) {
#     install.packages(pkg, repos="https://cloud.r-project.org/")
#   }
#   suppressPackageStartupMessages(library(pkg, character.only = TRUE))
# }

dunnett_power_analytic <- function(mu0, mu_t, n0, nt, sigma, alpha) {
  k  <- length(mu_t)
  ni <- rep(nt, k)
  se <- sqrt(1 / n0 + 1 / ni) * sigma
  ncp <- as.numeric((mu_t - mu0) / se)

  #  df = N_total - G,  G = k + 1
  df  <- n0 + sum(ni) - (k + 1)

  # corr((Ti-C),(Tj-C)) = (1 / n0) / sqrt((1 / ni + 1 / n0)(1 / nj + 1 / n0))
  R <- diag(k)
  for(i in 1:k){
    for(j in 1:k){
      if(i != j){
        R[i,j] <- (1 / n0) / sqrt((1 / ni[i] + 1 / n0) * (1 / ni[j] + 1 / n0))
      }
    }
  }

  # P(max |T_i| <= c) = 1 - alpha
  q <- qmvt(p = 1 - alpha, tail = "both.tails", df = df, corr = R)$quantile
  crit <- as.numeric(q)

  # per-comparison power i： 1 - [F(c; df, ncp) - F(-c; df, ncp)]
  pc_power <- 1 - (pt(crit, df, ncp) - pt(-crit, df, ncp))

  # familywise power：1 - P(\forall |T_i| <= c)
  keep_prob <- pmvt(lower = rep(-crit, k), upper = rep(crit, k),
                    df = df, corr = R, delta = ncp)
  fw_power <- 1 - as.numeric(keep_prob)

  list(
    df = df, crit = crit, 
    # R = R,
    per_comparison_power = pc_power,
    familywise_power = fw_power
  )
}

options(warn = 0)

File: doe_modules/simulation/_dunnett_power.py
import numpy as np
import pandas as pd
import rpy2.robjects as ro
import rpy2.robjects.vectors as ro_vectors
from rpy2.robjects.packages import importr, isinstalled
from rpy2.robjects import numpy2ri

from ._abstract import AbstractSimulator
from ._dunnett import Dunnett
from ._anova_power import sigma2


numpy2ri.activate()

utils = importr('utils')
base = importr('base')
packages = ["mvtnorm", "stats"]
repos = "https://cloud.r-project.org/"
rscript = f"{os.path.dirname(__file__)}/dunnett_power.R"
func_name = "dunnett_power_analytic"
R_FUNC = None


for pkg in packages:
    if not isinstalled(pkg): 
        utils.install_packages(pkg, repos=repos)
    base.suppressPackageStartupMessages(
        base.library(pkg, character_only=ro_vectors.BoolVector([True]))
    )


with open(rscript) as f:
    ro.r(f.read())


def _initialize_r_func():
    global R_FUNC
    if R_FUNC is None:
        R_FUNC = ro.r[func_name]


def dunnett_power(
    simulation: AbstractSimulator,
    alpha: float = 0.05,
    power_type: str = "per_comparison_power",
) -> np.ndarray:
    assert power_type in ["per_comparison_power", "familywise_power"], \
        f"power_type should be either 'per_comparison_power' or 'familywise_power', got {power_type}"
    dunnett_model = Dunnett(simulation)
    if dunnett_model.n_rep > 1:
        dunnett_model.summary()
        _initialize_r_func()
        power_dict = (
            lambda listvector: dict(zip(listvector.names, listvector))
        )(
            R_FUNC(
                mu0=float(dunnett_model.baseline.y),
                mu_t=dunnett_model.coef.y.values.ravel(),
                n0=dunnett_model.n_rep,
                nt=dunnett_model.n_rep,
                sigma=np.sqrt(sigma2(dunnett_model.simulation).item()),
                alpha=alpha
            )
        )

    else:
        power_dict =  {
            "df": np.nan,
            "crit": np.nan,
            "per_comparison_power": [np.nan] * (len(dunnett_model.group) - 1),
            "familywise_power": np.nan
        }

    return pd.DataFrame({
        "term": [v for v in dunnett_model.group if v != "all factors"],
        "power": power_dict[power_type],
        "model": ["Dunnett"] * (len(dunnett_model.group) - 1),
        "n_rep": [dunnett_model.n_rep] * (len(dunnett_model.group) - 1)
    })